File: server/src/main/java/com/ishop/ishoptech/models/product/Product.java
package com.ishop.ishoptech.models.product;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.ishop.ishoptech.models.BaseEntity;
import com.ishop.ishoptech.models.chipset.Chipset;
import com.ishop.ishoptech.models.company.Company;
import com.ishop.ishoptech.models.image.Image;
import com.ishop.ishoptech.models.receipt.ReceiptProduct;
import com.ishop.ishoptech.models.type.Type;
import lombok.*;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;

import javax.persistence.*;
import java.math.BigDecimal;
import java.util.List;
import java.util.Set;

@Data
@Table
@Entity(name = "product")
@AllArgsConstructor
@NoArgsConstructor
@With
@ToString
public class Product extends BaseEntity {

    private String name;
    private BigDecimal price;
    private String description;
    private String fullDescription;
    private Integer amount;

    @ToString.Exclude
    @ManyToOne(fetch = FetchType.LAZY)
    @Fetch(FetchMode.JOIN)
    @JoinColumn(name = "idType")
    private Type type;

    @ToString.Exclude
    @ManyToOne(fetch = FetchType.LAZY)
    @Fetch(FetchMode.JOIN)
    @JoinColumn(name = "idCompany")
    private Company company;

    @ToString.Exclude
    @ManyToOne(fetch = FetchType.LAZY)
    @Fetch(FetchMode.JOIN)
    @JoinColumn(name = "idChipset")
    private Chipset chipset;

    @JsonProperty(value = "images")
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @JoinColumn(name = "idProduct")
    @Fetch(FetchMode.JOIN)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private List<Image> images;

    @JsonProperty(value = "receiptProduct")
    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @JoinColumn(name = "idProduct")
    @Fetch(FetchMode.JOIN)
    @ToString.Exclude
    @EqualsAndHashCode.Exclude
    private Set<ReceiptProduct> receiptProducts;

}

File: client/shop/src/app/models/product/Product.ts
import {Type} from "../type/Type";
import {Company} from "../company/Company";
import {Chipset} from "../chipset/Chipset";
import {Image} from "../image/Image";

export class Product {
  id: number;
  name: string;
  price: bigint;
  description: string;
  amount: number;
  type: Type;
  company: Company;
  chipset: Chipset;
  images: Image[];

  constructor(id: number, name: string, price: bigint, description: string, amount: number, type: Type, company: Company, chipset: Chipset, images: Image[]) {
    this.id = id;
    this.name = name;
    this.price = price;
    this.description = description;
    this.amount = amount;
    this.type = type;
    this.company = company;
    this.chipset = chipset;
    this.images = images;
  }
}
